### PR TITLE
Fix/pdf ms case new data struct

### PR DESCRIPTION
**Pull request by dannil76**
## Explain the changes you’ve made
Added support for new case data structure.

## Explain why these changes are made
Without support for the new case data structure this lambda would fail and not generate a pdf as expected.

## Explain your solution
Parse case data from event.
Get cases that relates to the current user.
Sort cases based on date to get the two latest cases.
Generates a pdf based on the data from the case.
Saves the pdf as a base64 in s3 and updates the case attribute pdfGenerated to "yes".

## How to test the changes?

**Setting up your environment**
Setup a new secret in Systems Manager -> Paramter Store with the following information:
```
name: /pdfEnvs/dev,
type: SecureString,
value: {
  "recurringFormId": "INSERT_YOUR_RECURRING_FORM_ID", 
  "templatesFilenames": { 
     "EKB-recurring": { 
        "pdfBaseFilename": "ekb_recurring_base.pdf", 
        "JSONTemplateFilename": "ekb_recurring_template.json"
     }, 
     "default": { 
        "pdfBaseFilename": "ekb_recurring_base.pdf", 
        "JSONTemplateFilename": "ekb_recurring_template.json" 
     }
   } 
 }
 ```

**First scenario**
Make sure there are no cases connected to the user in DynamoDb table.
Setup an ongoing case for the user in your DynamoDb table.
Update the new case with form answers and set the case status too active:submitted:viva.

Result: Should generate a PDF from the case answers.

**Second scenario**
Setup a new ongoing case for a user in your DynamoDb cases-table.
Make sure you have a closed case in your DynamoDb cases-table ( case with a status type of closed:* ).
Update the ongoing case with answers and set the case status to active:submitted:viva.

Result: Should generate a PDF from the case answers with indications of changed values from the previous case.

DynamoDb case example data:
```
{
  // insert personalNumber in PK and SK to an existing user in your DynamoDB USER table.
  "PK": "USER#INSERT_PERSONAL_NUMBER",
  "SK": "USER#195809262743#CASE#INSERT_PERSONAL_NUMBER",
  "id": "9a939342-acc5-4357-bea4-4b470899ce25",
  "createdAt": 1612515760317,
  "updatedAt": 1612515760317
  "provider": "VIVA",
  "details": {
    "administrators": [],
    "period": {
      "end": 1614470400000,
      "start": 1612137600000
    },
    "workflowId": "459A8735121899A4C12586650047F916"
  },
  "currentFormId": "0",
  "forms": {
    "0": {
      "answers": [],
      "currentPosition": {
        "currentMainStep": 1,
        "currentMainStepIndex": 0,
        "index": 0,
        "level": 0
      }
    }
  },
  "status": {
    "type": "active:ongoing",
    "name": "Pågående",
    "description":
      "Du har påbörjat en ansökan. Du kan öppna din ansökan och fortsätta där du slutade.",
  },
  "pdfGenerated": "no" // Due to a bug we cannot set this to boolean since CloudFormation has a bugg that turns the expression into a string (true => "true") when querying.
}
```

### CODE REVIEW

File: services/pdf-ms/lambdas/addPdfToCase.ts [hunk 1]
```diff
@@ -1,138 +1,95 @@
...
+  const [currentCase, previousCase] = sortCasesByDate(userCases, 2);
+
+  const ssmParams = await PDF_SSM_PARAMS;
+  const currentCaseAnswers = currentCase.forms[ssmParams.recurringFormId].answers;
+  const previousCaseAnswers = previousCase.forms[ssmParams.recurringFormId].answers;
+
+  const { changedValues, newValues } = getNewAndChangedCaseAnswerValues(
+    currentCaseAnswers,
+    previousCaseAnswers
+  );
```
Comment (jwinding) [inline services/pdf-ms/lambdas/addPdfToCase.ts:52]: This breaks if the user only has a single case. In that case, the previousCase becomes undefined, and when accessing previousCase.forms[...] it crashes. So, we need to handle that, and decide if in this case we want everything to be colored red or everything colored black.


In general I find the logic here a little suspect. We simply get all cases on the user, but there's no guarantee that the cases will contain the relevant form, so even if the case exists, the previousCase.forms[ssmParams.recurringFormId] might not exist, so there needs to be some filtering on the cases to only pick those that contain the correct formId in forms.

File: services/pdf-ms/lambdas/addPdfToCase.ts [hunk 2]
```diff
@@ -1,138 +1,95 @@
...
-  const templates = JSON.parse((await loadFileFromBucket('templateFiles.json')).toString());
-  const templateFiles = templates[Object.keys(templates).includes(formType) ? formType : 'default'];
+  const pdfJsonValues = {
+    ...arrayToObject(currentCaseAnswers),
+    user,
+    valuesChanged: changedValues.length > 0 || newValues.length > 0 ? 'Ja' : 'Nej',
+  };
+
+  const templates = ssmParams.templatesFilenames;
+  const ekbTemplateFiles = templates['EKB-recurring'];
```
Comment (jwinding) [inline services/pdf-ms/lambdas/addPdfToCase.ts:67]: hardcoding this to only work for EKB-recurring feels bad, but okay for now, I guess...

**Review by jwinding — DISMISSED** [review]
A lot to go through here, but for starters I've found a bug, that was probably already here but that needs fixing. The case where the user only has a single case for the form is not handled correctly, the service crashes, see my comment.

**Comment by TeddyAndersson** [comment]
The following test scenarios have been tested and works:

Creating a pdf when there is no previous closed cases. 

Creating a pdf when there is one previous closed case.

Creating a pdf when there is several previous closed cases. Creates a pdf and check values of the most recently closed case.

Creating a pdf on a user that does not exists. You cannot create a pdf on a user that does not exists.

Theses test have been made from the AWS without app involvement. So further testing with the app will need to be done.

**Review by TeddyAndersson — APPROVED** [review]
These changes have been tested and works on AWS. Therefore, there are minor changes that will need to be addressed in order to make the generated pdf more understandable. But these changes are old issues and do not impact this PR.

New task for correcting the pdf template and how it renders should be added to the backlog.

**Comment by zilaei** [comment]
Missing Swish in the PDF. 
I'm not in the loop so this might be by something for the next version: Not seeing the imputed Swish amount in the PDF.
Testing through the app. This is the case result, with generated PDF / missing Swish:

[case_no_swish.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6077172/case_no_swish.txt)

**Comment by zilaei** [comment]
> > Missing Swish in the PDF.
> > I'm not in the loop so this might be by something for the next version: Not seeing the imputed Swish amount in the PDF.
> > Testing through the app. This is the case result, with generated PDF / missing Swish:
> > [case_no_swish.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6077172/case_no_swish.txt)
> 
> Realize that I might have screwed up with swish IDs for the form. I will double check and try again.

Tried again from scratch without success, did not see imputed Swish value in PDF.
Replicate by following steps:
1) Cleared all previous cases from personal AWS DB.
2) Copied form "Ansökan för ekonomiskt bistånd - löpande - Den mest aktuella inför release 1" from Mitt HBG Develop Develop AWS:
[form_Ansökan_för_ekonomiskt_bistånd.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078497/form_Ansokan_for_ekonomiskt_bistand.txt)
3) Stepped through form and filled inputs. Result of case, after submitted can be found in file:
[case_result.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078533/case_result.txt)
4) Swish can not be found in PDF. Copy of generated PDF:
[c414f388-4f81-4c3a-b3ca-d9874a8df520.pdf](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078547/c414f388-4f81-4c3a-b3ca-d9874a8df520.pdf)

**Comment by zilaei** [comment]
> > > Missing Swish in the PDF.
> > > I'm not in the loop so this might be by something for the next version: Not seeing the imputed Swish amount in the PDF.
> > > Testing through the app. This is the case result, with generated PDF / missing Swish:
> > > [case_no_swish.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6077172/case_no_swish.txt)
> > 
> > 
> > Realize that I might have screwed up with swish IDs for the form. I will double check and try again.
> 
> Tried again from scratch without success, did not see imputed Swish value in PDF.
> Replicate by following steps:
> 
>     1. Cleared all previous cases from personal AWS DB.
> 
>     2. Copied form "Ansökan för ekonomiskt bistånd - löpande - Den mest aktuella inför release 1" from Mitt HBG Develop Develop AWS:
>        [form_Ansökan_för_ekonomiskt_bistånd.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078497/form_Ansokan_for_ekonomiskt_bistand.txt)
> 
>     3. Stepped through form and filled inputs. Result of case, after submitted can be found in file:
>        [case_result.txt](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078533/case_result.txt)
> 
>     4. Swish can not be found in PDF. Copy of generated PDF:
>        [c414f388-4f81-4c3a-b3ca-d9874a8df520.pdf](https://github.com/helsingborg-stad/helsingborg-io-sls-api/files/6078547/c414f388-4f81-4c3a-b3ca-d9874a8df520.pdf)

This issue seemed to originate from tags in the form and something that existed before this code: Issue is out of scope so this PR and code works! Good job!

**Review by zilaei — APPROVED** [review]

**Pull request merged**